### ISSUE TITLE
operations: let's use threads for collecting operation files

### ISSUE THREAD

**Issue by zhucan** [rclone/rclone#4808]
https://github.com/rclone/rclone/blob/979bb07c86f9680466e9421da116aaf3af94c1da/fs/march/march.go#L418

@ncw I think we can do it (NewObject) with multi threads.

**Comment by ivandeex** [comment]
Feel free to submit a PR.

**Comment by ncw** [comment]
> I think we can do it (NewObject) with multi threads.

Yes that would be a good idea.

Do you want to make a PR for this @zhucan ?

You'll need a concurrency limiter and a loop something like this

```go
        tokens := make(chan struct{}, fs.Config.Checkers)
        var wg sync.WaitGroup
        for /*...*/{
                tokens <- struct{}{}
                wg.Add(1)
                go func(stuff type) {
                        defer func() {
                            wg.Done()
                            <-tokens
                        }()
                       // stuff
                }(stuff)
        }
        wg.Wait()
```

You'll also need a `sync.Mutex` so you hold it when appending to the output list.

**Comment by ivandeex** [comment]
Related to https://github.com/rclone/rclone/issues/4798 (multi-thread upload interface)

**Comment by zhucan** [comment]
Ok, I will do it out of day发自我的iPhone------------------ Original ------------------From: "Ivan Andreev" <notifications@github.com> <"Ivan Andreev" <notifications@github.com>>Date: Thu,Nov 26,2020 5:19 PMTo: zhucan.k8s <zhucan.k8s@gmail.com>Subject: Re: [rclone/rclone] 【question】why not do it with multi threads? (#4808)
Feel free to submit a PR.

—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or unsubscribe.
[
{
"@context": "http://schema.org",
"@type": "EmailMessage",
"potentialAction": {
"@type": "ViewAction",
"target": "https://github.com/rclone/rclone/issues/4808#issuecomment-734175082",
"url": "https://github.com/rclone/rclone/issues/4808#issuecomment-734175082",
"name": "View Issue"
},
"description": "View this Issue on GitHub",
"publisher": {
"@type": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

**Comment by ivandeex** [comment]
Resolved by #4821
